cmake/extensions.cmake

```diff
@@ -930,11 +945,42 @@ function(zephyr_linker_sources location)
       message(FATAL_ERROR "zephyr_linker_sources() was called on a directory")
     endif()
 
-    # Append the file contents to the relevant destination file.
-    file(READ ${path} snippet)
-    file(RELATIVE_PATH relpath ${ZEPHYR_BASE} ${path})
-    file(APPEND ${snippet_path}
-             "\n/* From \${ZEPHYR_BASE}/${relpath}: */\n" "${snippet}\n")
+    # Find the relative path to the linker file from the include folder.
+    file(RELATIVE_PATH relpath ${ZEPHYR_BASE}/include ${path})
+
+    # Create strings to be written into the file
+    set (key_str "/* Key: \"${KEY}\" */")
+    set (include_str "#include \"${relpath}\"")
+
+    file(STRINGS ${snippet_path} lines) # Get current lines (without newlines).
+    list(FIND lines ${key_str} key_index) # Find key if it already exists.
+    if (${key_index} EQUAL -1)
+      # Key doesn't exist in file yet.
+      file(STRINGS ${snippet_path} key_lines REGEX "Key: ") # Get just key lines
+      set(key_lines ${key_lines} ${key_str}) # Add new key line
+      list(SORT key_lines)
+      list(FIND key_lines ${key_str} new_key_index)
+      list(LENGTH key_lines key_lines_len)
+      MATH(EXPR key_lines_last "${key_lines_len}-1")
+      if (${key_lines_last} EQUAL ${new_key_index})
+        # New key is last.
+        list(APPEND lines ${key_str})
+        list(FIND lines ${key_str} key_index)
+      else()
+        # New key not last, place it correctly in ${lines}.
+        MATH(EXPR after_new_key_index "${new_key_index}+1")
+        list(GET key_lines ${after_new_key_index} after_new_key)
+        list(FIND lines ${after_new_key} key_index)
+        list(INSERT lines ${key_index} ${key_str})
+      endif()
+    endif()
+
+    # Place include string after its key.
+    MATH(EXPR include_index "${key_index}+1")
+    list(INSERT lines ${include_index} ${include_str})
+
+    string(REPLACE ";" "\n;" lines "${lines}") # Add newline to each line.
+    file(WRITE ${snippet_path} ${lines} "\n")
```

Comment:
I believe we can simplify this if we append

`/* Key: 0x0 */ #include <zephyr.h>
`

to the file and then sort.

According to my tests and to a certain degree the gcc documentation and C spec, it is an error to put a comment after an #include, but not to put a comment before an #include.

```
/* Key: 0x0 */ #include <zephyr.h>
/* Key: default */ #include <sys/printk.h>
/* Key: default */ #include /* Translation phase 3 removes this before '#' is interpreted in phase 4, so this is OK. */ <sys/printk.h>

```

> A preprocessing directive consists of a sequence of preprocessing tokens that begins with
> a # preprocessing token that (at the start of translation phase 4) is either the first character
> in the source file (optionally after white space containing no new-line characters) or that
> follows white space containing at least one new-line character, and is ended by the next
> new-line character.140)

(Comments are removed at phase 3, so we are good for all STD compliant toolchains AFAICT).

Reply:
This looks really clever, I'll try it out. Thanks!